anker/src/state.rs

```diff
@@ -42,15 +52,147 @@ impl Anker {
+        }
+        Ok(())
+    }
+
+    /// Confirm that the derived account address matches the `account_info` adddress.
+    fn check_derived_account_address(
+        &self,
+        name: &'static str,
+        seed: &'static [u8],
+        bump_seed: u8,
+        anker_program_id: &Pubkey,
+        anker_instance: &Pubkey,
+        account_info: &AccountInfo,
+    ) -> ProgramResult {
+        let address = Pubkey::create_program_address(
+            &[anker_instance.as_ref(), seed, &[bump_seed]],
+            anker_program_id,
+        )
+        .expect("Depends only on Anker-controlled values, should not fail.");
+
+        if *account_info.key != address {
+            msg!(
+                "Expected {} to be {}, but found {} instead.",
+                name,
+                address,
+                account_info.key,
+            );
+            return Err(AnkerError::InvalidDerivedAccount.into());
+        }
+        Ok(())
+    }
+
+    /// Confirm that the provided reserve account is the one that belongs to this instance.
+    ///
+    /// This does not check that the reserve is an stSOL account.
+    pub fn check_reserve_address(
+        &self,
+        anker_program_id: &Pubkey,
+        anker_instance: &Pubkey,
+        reserve_account_info: &AccountInfo,
+    ) -> ProgramResult {
+        self.check_derived_account_address(
+            "the reserve account",
+            ANKER_RESERVE_ACCOUNT,
+            self.reserve_account_bump_seed,
+            anker_program_id,
+            anker_instance,
+            reserve_account_info,
+        )
+    }
+
+    /// Confirm that the provided reserve authority is the one that belongs to this instance.
+    pub fn check_reserve_authority(
+        &self,
+        anker_program_id: &Pubkey,
+        anker_instance: &Pubkey,
+        reserve_authority_info: &AccountInfo,
+    ) -> ProgramResult {
+        self.check_derived_account_address(
+            "the reserve authority",
+            ANKER_RESERVE_AUTHORITY,
+            self.reserve_authority_bump_seed,
+            anker_program_id,
+            anker_instance,
+            reserve_authority_info,
+        )
+    }
+
+    /// Confirm that the provided bSOL mint authority is the one that belongs to this instance.
+    pub fn check_mint_authority(
+        &self,
+        anker_program_id: &Pubkey,
+        anker_instance: &Pubkey,
+        mint_authority_info: &AccountInfo,
+    ) -> ProgramResult {
+        self.check_derived_account_address(
+            "the bSOL mint authority",
+            ANKER_MINT_AUTHORITY,
+            self.mint_authority_bump_seed,
+            anker_program_id,
+            anker_instance,
+            mint_authority_info,
+        )
+    }
+
+    /// Confirm that the provided mint account is the one stored in this instance.
+    pub fn check_mint(&self, provided_mint: &AccountInfo) -> ProgramResult {
+        if *provided_mint.owner != spl_token::id() {
```

Comment:
After thinking about it for a bit, I think in this case it would be safe to not check the owner (because the mint is controlled by us, and we won’t delete it and put a different account there after initialization). But I feel safer checking it anyway.